core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -115,6 +116,21 @@ public Optional<Rel> visit(Join join) throws RuntimeException {
             .build());
   }
 
+  @Override
+  public Optional<Rel> visit(Set set) throws RuntimeException {
+    var visitedInputs = set.getInputs().stream().map(inputRel -> inputRel.accept(this)).toList();
+    if (visitedInputs.stream().noneMatch(visitedInput -> visitedInput.isPresent())) {
+      return Optional.empty();
+    }
+    // replace all optional values with the input values
+    var setInputs =
+        IntStream.range(0, set.getInputs().size())
+            .mapToObj(i -> visitedInputs.get(i).orElse(set.getInputs().get(i)))
+            .toList();
+    return Optional.of(
+        ImmutableSet.builder().from(set).inputs(setInputs).setOp(set.getSetOp()).build());
+  }
+
```

Comment:
Since all other traversals of expressions and relations are already going through transformList, I'd be inclined to have this one do the same. Maybe worthwhile to rename transformList or add a comment? Here's what it would look like using that function:
```
    return transformList(set.getInputs(), t -> t.accept(this))
        .map(u -> ImmutableSet.builder().from(set).inputs(u).setOp(set.getSetOp()).build());

```

Reply:
thanks for pointing this out, it's indeed better to re-use the transformList function, I missed it by mistake.